Extensibility: Plugin sidebar api

abotteram opened the pull request:
## Description
<!-- Please describe your changes -->
Also props to @IreneStr.

Usage:
```js
const { Fragment } = wp.element;
const { PluginSidebar, registerPlugin } = wp.editPost;

// Separate component to avoid "unnamed component" warning.
const Component = () => {
	return (
		<Fragment>
			<PluginSidebar name="sidebar" title="Sidebar title">
				<h2>Some content!</h2>
			</PluginSidebar>
			<PluginSidebar name="second-sidebar" title="Sidebar 2 title">
				<h2>Some more content!</h2>
			</PluginSidebar>
		</Fragment>
	);
};

// Register the plugin using the new API
registerPlugin( {
	name: "plugin-namespace", // Should start with a lowercase letter and can only contain lowercase letters, numbers and dashes.
	render: Component,
} );
```

**Important:** Due to a bug in the current API you should open and close the editor sidebar before executing the following line.

Then to activate the sidebar execute the following line in the console:
`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin', 'plugin-namespace/sidebar' );`
or
`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin', 'plugin-namespace/second-sidebar' );`

Known issues:
* Providing an unknown namespace/sidebarName id opens a white sidebar.
* A way to activate the sidebar without needing the console.

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
* Tested in editor, using console.
* Added unit tests.

## Screenshots (jpeg or gifs if applicable):

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

Commit:
Initial implementation of slot-fill sidebar API

Commit:
Initial implementation of slot-fill sidebar API

Commit:
Attempt to implement createContext api for plugin slot fill API

Commit:
Attempt to implement createContext api for plugin slot fill API

Commit:
Fixed error on start up

Commit:
Fixed error on start up

gziolo commented on edit-post/components/plugins-panel/index.js:
```diff
@@ -44,7 +45,7 @@ function PluginsPanel( { onClose, plugin } ) {
 				/>
 			</div>
 			<div className="edit-post-plugins-panel__content">
```
It seem like `PluginsPanel` should be converted into the slot.

Xyfi replied:
Yes, that's the plan.

gziolo commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,97 @@
+					const Context = plugin.context;
+					return (
+						<Context.Provider key={ plugin.name } value={ plugin.name }>
+							{ plugin.render() }
```
Shouldn't you wrap this with `<PluginSidebar />` to activate the fill?

Xyfi replied:
`PluginSidebar` is the component the plugin developer uses in the `registerPlugin` API call. When rendering the plugins you don't know what kind of ui-element it is (Sidebar, screen takeover, etc).

gziolo commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,97 @@
+				{ map( plugins, plugin => {
+					const Context = plugin.context;
+					return (
+						<Context.Provider key={ plugin.name } value={ plugin.name }>
```
This is also a good place to put logic to make sure that sidebar is only rendered when a given plugin is active.

Commit:
Implemented way to pass props from plugin API via context to PluginSidebar component

Commit:
Implemented way to pass props from plugin API via context to PluginSidebar component

Commit:
Finalize first pass of registerPlugin API

Commit:
Finalize first pass of registerPlugin API

Commit:
Remove create-react-context

Commit:
Remove create-react-context

Commit:
Merge branch 'master' into add/slot-fill-sidebar-api

Commit:
Removed old sidebar and refactored the way plugin context is handled

Commit:
Removed old sidebar and refactored the way plugin context is handled

Commit:
Improved withPluginContext HOC to prevent usage of cloneElement

Commit:
Improved withPluginContext HOC to prevent usage of cloneElement

Commit:
Implemented tests for registerPlugin

Commit:
Implemented tests for registerPlugin

Commit:
Added test for plugin/context components, revamped naming of passed context

Commit:
Added test for plugin/context components, revamped naming of passed context

Commit:
Fixed eslint issues

Commit:
Fixed eslint issues

gziolo commented on edit-post/api/index.js:
```diff
@@ -1,4 +1,3 @@
-	registerSidebar as __experimentalRegisterSidebar,
-	activateSidebar as __experimentalActivateSidebar,
-} from './sidebar';
+	registerPlugin,
```
Nice 👍

gziolo replied:
To repeat what I said above. Let's mark it as experimental for one release.

gziolo commented on edit-post/api/components/test/context.js:
```diff
@@ -0,0 +1,33 @@
+
+			const tree = renderer.toJSON();
+
+			expect( tree.children[ 0 ] ).toEqual( 'plugin-namespace' );
```
Gutenberg has [jest-enzyme](https://www.npmjs.com/package/jest-enzyme) bundled, it is possible to use snapshots here and `mount` helper. See: https://github.com/WordPress/gutenberg/blob/master/docs/testing-overview.md#snapshot-testing.

Commit:
Refactored plugin-sidebar to use withSelect/withDispatch and created separate sidebar layout component

Commit:
Refactored plugin-sidebar to use withSelect/withDispatch and created separate sidebar layout component

Commit:
Removed PropTypes

Commit:
Removed PropTypes

gziolo commented on edit-post/components/layout/index.js:
```diff
@@ -35,21 +35,20 @@ import {
 			return <Sidebar />;
 		case 'plugin':
-			return <PluginsPanel />;
+			return <PluginSidebarSlot />;
```
It would be nice to avoid using the `slot` part in the name of the component. It should be known only internally.

gziolo commented on edit-post/components/layout/index.js:
```diff
@@ -111,6 +107,7 @@ function Layout( {
 					openedGeneralSidebar={ openedGeneralSidebar } />
 			}
 			<Popover.Slot />
+			<PluginFills />
```
This component might be renamed to `PluginArea` or something like this to better reflect its role.

gziolo commented on edit-post/components/plugin-sidebar/index.js:
```diff
@@ -0,0 +1,98 @@
+ */
+const SLOT_NAME = 'PluginSidebar';
+
+class SidebarErrorBoundary extends Component {
```
This component should be put in its own file. In the long run, it should be generalized to support or types of plugins.

gziolo commented on edit-post/components/plugin-sidebar/index.js:
```diff
@@ -0,0 +1,98 @@
+	constructor( props ) {
+		super( props );
+
+		if ( typeof props.name !== 'string' ) {
```
Do we need to include this validation in here? I don't feel like anything bad would happen if the non-string value would be provided. It would be converted to a string and it wouldn't be possible to activate such sidebar.

Xyfi replied:
It's a matter of providing accurate feedback to the developer I think?

gziolo replied:
It feels like this kind of behavior should be enabled only in `development` env. I don't know, @mtias and @aduth - do we have some strategy in place?

aduth replied:
We shouldn't throw.

https://github.com/WordPress/gutenberg/pull/5430#discussion_r173636296

Commit:
Used enzyme instead of react-test-renderer in plugin context tests

Commit:
Used enzyme instead of react-test-renderer in plugin context tests

Commit:
Hidden internal Slot and Fill components in layout

Commit:
Hidden internal Slot and Fill components in layout

Commit:
Moved plugin-sidebar error-boundary to separate file

Commit:
Moved plugin-sidebar error-boundary to separate file

Commit:
Moved sidebar error boundary around the component only so the error can be rendered within the sidebar

Commit:
Moved sidebar error boundary around the component only so the error can be rendered within the sidebar

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Updated documentation

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,65 @@ wp.hooks.addFilter(
+### Available UI components
 
-**Example**
+The available UI components are available under `wp.editPost`, and are React components.
```
available is repeated in this sentence, it might be a good idea to rephrase it.

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,65 @@ wp.hooks.addFilter(
+
+`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin-name/sidebar-name' );`
+
+Notice that you need to use both the plugin name and sidebar name separated by a `/` to show the correct sidebar.
```
I’m still wondering if it isn’t confusing that we register using 2 separate parts but here magically join using slash. We should consider having user facing API that better reflects the way we register plugins.

Xyfi replied:
I'm not really sure how to do it otherwise. This is the only to make sure the sidebar is unique.

Any suggestions?

Commit:
Updated documentation

Commit:
Updated documentation

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
-**Warning:** This is an experimental API, and is subject to change or even removal.
-
-### Registering a sidebar
+## Extending the editor's UI (Slot and Fill)
```
I think we should keep it experimental for 2.4 until we have everything better tested.

> ## Extending the editor's UI (Experimantal)

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
+Extending the editor's UI is done with in `registerPlugin` API, and allows you to define all your plugin's UI elements in one place.
 
-This method takes a sidebar `name` and a `settings` object, containing a title and a render function. The name should contain a namespace prefix (Example: my-plugin/my-custom-sidebar).
+`wp.editPost.registerPlugin( { name: string, render: function } )`
```
> `wp.editPost.__experimentalRegisterPlugin( { name: string, render: function } )`

gziolo replied:
It might be a good idea to make it a header in a similar way we do it in `data` module:

https://github.com/WordPress/gutenberg/blob/master/data/README.md#select-storename-string--object

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
+
+const Component = () => (
+	<Fragment>
+        <PluginSidebar name="first-sidebar-name" title="My Sidebar">
```
There are some issues with the spacing here when you check the preview:

![screen shot 2018-03-08 at 15 01 48](https://user-images.githubusercontent.com/699132/37154902-a9a45974-22e1-11e8-9a38-c896853f3a79.png)

gziolo replied:
I think it is also confusing that we use h1 in here, it might give an impression that it should be a header of the plugin, but it isn't.

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
 
-`wp.editPost.__experimentalActivateSidebar( name: string )`
+`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin-name/first-sidebar-name' );`
+`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin-name/second-sidebar-name' );`
```
![screen shot 2018-03-08 at 15 08 44](https://user-images.githubusercontent.com/699132/37155227-a31a8b22-22e2-11e8-8786-ee9c22a19e59.png)
Not sure why, but they display in one line...
It might help to treat it as a code snippet.

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
+### Available UI components
 
-**Example**
+The available UI components are found in the global variable `wp.editPost`, and are React components.
```
> and are React components.

I think we refer to them as `WordPress Components`. See: 
https://github.com/WordPress/gutenberg/blob/master/element/index.js#L44-L47

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
-## Extending the editor's UI (Slot and Fill)
+Renders a sidebar when activated.
+
+`<PluginSidebar name="sidebar-name" title="Sidebar title">{ contents }</PluginSidebar>`
```
This should be converted to code snippet, too. Might look as follows:
```js
<PluginSidebar
	name="sidebar-name"
	title="Sidebar title"
>
	{ contents }
</PluginSidebar>

gziolo commented on docs/extensibility.md:
````diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
 ```
 
-### Activating a sidebar
+You can activate the sidebars using the following lines:
````
I think we can remove lines 157-160 as we have a similar thing put later in a better-scoped context.

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,64 @@ wp.hooks.addFilter(
+
+The sidebar can be activated using the data api:
+
+`wp.data.dispatch( 'core/edit-post' ).openGeneralSidebar( 'plugin-name/sidebar-name' );`
```
Let's convert it to the code snippet.

gziolo commented on edit-post/api/README.md:
```diff
@@ -3,35 +3,60 @@ Edit post API
 The edit post API contains the following methods:
 
-### `wp.editPost.__experimentalRegisterSidebar( name: string, settings: { title: string, render: function } )`
+## `wp.editPost.registerPlugin( { name: string, render: function } )`
```
I think it is copy and paste, so the same feedback as above :)

gziolo commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,87 @@
+		return null;
+	}
+
+	settings = applyFilters( 'editPost.registerPlugin', settings, settings.name );
```
I don't think it is very useful at the moment as there are only 2 properties `name` and `render`. We can leave it you think that there is value in it, but it should be moved to the beginning of the plugin. I'm also not sure if changing the name of the plugin is a good idea because it might break activation logic.

Xyfi replied:
@aduth set the opposite in a previous PR, that the filters should be applied after the plugin has been validated.

gziolo replied:
You can override those properties to any random value in that case. Why wpuld we care only about the original params passed but not about those set after all filters got applied?

Xyfi replied:
Because a plugin should at the very least be valid? :)

gziolo replied:
Try this:
```js
wp.hooks.addFilter( 'editPost.registerPlugin', 'my-unique-name', function() {
    return 'Whatever;
} );
```
and let me know if it is going to work properly with your plugin :)

aduth replied:
>@aduth said the opposite in a previous PR, that the filters should be applied after the plugin has been validated.

I did?

gziolo commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,87 @@
+	render() {
+		return (
+			<div id="plugin-fills" style={ { display: 'none' } }>
+				{ map( plugins, plugin => {
```
I don't think it's a good idea to have it like this in the long run. It should be provided as a prop so the plugin area could rerender when a new plugin is registered. At the moment it won't rerender until it is going to be explicitly unmounted by the top component. It shouldn't be an issue for the time being as most of the plugins will register plugins before `edit-post` is loaded by the browser. We will have to fix it somehow.

Xyfi replied:
The most obvious choice would be to register plugins to the store and have this component listen to it.

gziolo replied:
Yes, using the store seems like the best fit. My only concern is that plugin object contains non-serializable parts - in particular, the component. I don't know how strict about that we want to be. In fact, I don't think we would ever want to persist the list of registered plugins. We might give it a try. We could move `PluginArea` to its own file with that in place.

gziolo commented on edit-post/components/layout/index.js:
```diff
@@ -62,15 +61,12 @@ function Layout( {
-		( openedGeneralSidebar !== 'plugin' || getSidebarSettings( plugin ) );
 	const className = classnames( 'edit-post-layout', {
-		'is-sidebar-opened': isSidebarOpened,
+		'is-sidebar-opened': layoutHasOpenSidebar,
```
> Providing an unknown namespace/sidebarName id opens a white sidebar.

This is the root cause. We will have to find some solution for it.

gziolo commented on edit-post/components/plugin-sidebar/error-boundary.js:
```diff
@@ -0,0 +1,27 @@
+import { Component } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+
+class SidebarErrorBoundary extends Component {
```
It should be called `PluginSidebarErrorBoundary` and the `className` used below should looks as follows: `plugin-sidebar__error-boundary`. See the guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

gziolo commented on edit-post/components/plugin-sidebar/style.scss:
```diff
@@ -18,4 +18,8 @@
   }
 }
 
+.plugin-sidebar-error {
```
I think this is a very basic support for plugin errors :)

@karmatosed @jasmussen @mtias  - do we have any plan how to handle errors for plugins? Should we provide a fallback for both `development` and `production`? Should the differ?

jasmussen replied:
Not sure, it would definitely be nice with a unified pattern, but that also has technical questions I'm not skilled enough to have insight in.

gziolo replied:
I will share screens later :)

gziolo commented:
I left my feedback. We are definitely on track, this is exactly what we discussed in #4484. I still need to catch with the Context API. I only played with the new shiny API which we can't use yet. We need to polish a few things and should be heading towards merging this one.

Commit:
Refactored editpost/api/context component to reparate folder

Commit:
Refactored editpost/api/context component to reparate folder

Commit:
Moved plugin components to api folder and made API experimental

Commit:
Moved plugin components to api folder and made API experimental

Commit:
Updated docs to reflect experimental status

Commit:
Updated docs to reflect experimental status

Commit:
Updated documentation

Commit:
Updated documentation

Commit:
Fixed failing test

Commit:
Fixed failing test

gziolo committed:
Docs: Add a few tweaks to the Extensibility doc

gziolo committed:
Docs: Add a few tweaks to the Extensibility doc

gziolo commented:
I added a few tweaks to the Extensibility doc with https://github.com/WordPress/gutenberg/pull/5430/commits/ca63d6266b46a5ad7943a98f8023acf45394a802 to speed it up. Feel free to add your tweaks and apply to the 2nd document. At some point, this will probably diverge which is fine.

aduth commented on edit-post/api/components/plugin-sidebar/style.scss:
```diff
@@ -18,4 +18,8 @@
 
+.plugin-sidebar-error {
+  padding: 16px;
+  color: red;
```
Do we have a better color we should be using than "red"? Or maybe not red at all? What's the design basis for this color choice?

aduth commented on edit-post/api/index.js:
```diff
@@ -1,4 +1,11 @@
+import { registerPlugin } from './plugin';
+import { PluginSidebarFill as PluginSidebar } from './components/plugin-sidebar';
```
I'm quite confused about this.

In `plugin-sidebar/index.js`:

```js
export { PluginSidebarFill, PluginSidebarSlot as PluginSidebar };
```

In `edit-post/api/index.js`:

```js
import { PluginSidebarFill as PluginSidebar } from './components/plugin-sidebar';
```

What is "PluginSidebar" ? We're using it in two different ways.

aduth replied:
Further: Elsewhere we define slot as properties of the class. We should be consistent one way or the other.

See:

https://github.com/WordPress/gutenberg/blob/8a4d644a525c1463f469e525d5d31ad36f44e23c/components/popover/index.js#L332

Xyfi replied:
Because I didn't want ot expose the slot to the plugin developer, but it can't hurt I guess.

aduth commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,87 @@
+class PluginArea extends Component {
+	render() {
+		return (
+			<div id="plugin-fills" style={ { display: 'none' } }>
```
Why `id` ?

Typically we avoid it if at all possible in components, because we don't want to guarantee that there's a single instance of it.

And where we do need one, we should use [`withInstanceId`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-instance-id) to help guarantee uniqueness of the ID.

aduth commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,87 @@
+/**
+ * A component that renders all plugin fills in a hidden div.
+ */
+class PluginArea extends Component {
```
Why are some components defined in a `components` directory, but this one here?

aduth commented on edit-post/api/index.js:
```diff
@@ -1,4 +1,11 @@
+};
+
 export {
-	registerSidebar as __experimentalRegisterSidebar,
```
While it is clearly flagged as `__experimental`, is there any way we can [`deprecated`](https://github.com/WordPress/gutenberg/blob/master/utils/deprecation.js) this easily with  a transparent upgrade path?

aduth commented on docs/extensibility.md:
```diff
@@ -120,38 +120,59 @@ wp.hooks.addFilter(
-		return <p>This is an example</p>;
-	},
+const Fragment = wp.element.Fragment;
+const PluginSidebar = wp.editPost.__experimental.PluginSidebar;
```
In fact, to the latter point, you could destructure `registerPlugin` to use it in the below example:

```js
const { PluginSidebar, registerPlugin } = wp.editPost.__experimental;

// ...

registerPlugin( {
	name: 'my-plugin-name',
	render: MyPluginComponent,
} );
```

aduth commented on edit-post/api/components/plugin-sidebar/index.js:
```diff
@@ -0,0 +1,90 @@
+					title={ props.title }
+					onClose={ props.onClose } >
+					<ErrorBoundary pluginName={ this.namespacedName }>
+						{ cloneElement( Children.only( children ), newProps ) }
```
`Children.only` is not accommodating to different types of elements. This will error when rendering a string:

```js
const { createElement } = wp.element;
const { registerPlugin, PluginSidebar } = wp.editPost.__experimental;

function MyPlugin() {
	return createElement(
		PluginSidebar,
		{ name: 'first-sidebar' },
		'My sidebar',
	);
}

registerPlugin( {
	name: 'demo-sidebar',

	render: MyPlugin,
} );
```

aduth replied:
Then, when rendering as an element, it will render, but with warnings, because we're passing props (why are we passing props?):

```js
const { createElement } = wp.element;
const { registerPlugin, PluginSidebar } = wp.editPost.__experimental;

function MyPlugin() {
	return createElement(
		PluginSidebar,
		{ name: 'first-sidebar' },
		createElement( 'p', null, 'My sidebar' )
	);
}

registerPlugin( {
	name: 'demo-sidebar',

	render: MyPlugin,
} );
```

aduth commented on edit-post/api/plugin.js:
```diff
@@ -0,0 +1,87 @@
+ *
+ * @return {Object} The final plugin settings object.
+ */
+function registerPlugin( settings ) {
```
Why is `registerPlugin` in `edit-post`? Will this pattern or the concept of plugin UI extensibility not apply to other screens?

gziolo replied:
See: https://github.com/WordPress/gutenberg/pull/5430#issuecomment-372607425. It makes perfect sense to introduce new `plugins` module to align with `blocks`.

aduth commented on edit-post/api/components/plugin-sidebar/sidebar-layout.js:
```diff
@@ -0,0 +1,31 @@
+const SidebarLayout = ( { onClose, title, children } ) => {
+	return (
+		<div
+			className="edit-post-sidebar edit-post-plugins-panel"
```
https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

aduth replied:
If we need to inherit styles from another component, we shouldn't simply assign the same class, we should create a separate component which is generalized to support both its current other usage, and the usage we're wanting here.

aduth commented on edit-post/api/components/plugin-sidebar/sidebar-layout.js:
```diff
@@ -0,0 +1,31 @@
+import './style.scss';
```
Move into "Internal dependencies" DocBlock below WordPress dependencies.

https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#imports

aduth commented on edit-post/api/index.js:
```diff
@@ -1,4 +1,11 @@
+import { registerPlugin } from './plugin';
```
DocBlock comment: https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#imports

Commit:
Rebase master

Commit:
Merged remote changes

Commit:
Merge branch 'master' of https://github.com/WordPress/gutenberg into add/slot-fill-sidebar-api

Commit:
Register UI components in plugin.js

gziolo commented:
Related discussion on Slack: https://wordpress.slack.com/archives/C02QB2JS7/p1520544138000624:
> Yes, it makes sense to group those features together. I’m still not sure how the final API and code structure should look like. Mayby we should have wp.plugins.registerPlugin and wp.plugins.PluginsArea but wp.editPost.PluginSidebar and wp.editPost.PluginMenuItem?

Just wanted to reshare what I said on Slack last week regarding code organization to second what @aduth commented above.

gziolo commented on edit-post/api/README.md:
```diff
@@ -3,35 +3,57 @@ Edit post API
 The edit post API contains the following methods:
 
-### `wp.editPost.__experimentalRegisterSidebar( name: string, settings: { title: string, render: function } )`
+### `wp.editPost.__experimental.registerPlugin( { name: string, render: function } )`
```
This is a huge limitation of our docs that there is no easy way to keep some parts in sync. It would be nice to come up with a solution. Maybe we should keep some docs next to the code but also expose them as part of the handbook?

gziolo commented on docs/extensibility.md:
```diff
@@ -120,38 +120,59 @@ wp.hooks.addFilter(
-`wp.editPost.__experimentalRegisterSidebar( name: string, settings: { title: string, render: function } )`
-
-This method takes a sidebar `name` and a `settings` object, containing a title and a render function. The name should contain a namespace prefix (Example: my-plugin/my-custom-sidebar).
+### `wp.editPost.__experimental.registerPlugin( { name: string, render: function } )`
```
I missed that. Yes, it might make sense to have `name` and `settings` here, assuming we will have some more settings in the future 😃

Commit:
Updated tests

Commit:
created PluginRegistry singleton

Commit:
Updated JSDoc

Commit:
Made name first parameter of registerPlugin function instead of being part of teh settings object

Commit:
Small documentation tweaks

Commit:
Moved and documented plugin context provider and higher order component

Commit:
Updated test

Commit:
Fixed sass class names

Commit:
Made scss specific for plugin0-sidebar which nog longer depends on edit-post-sidebar scss

Commit:
Moved PLuginsArea to separate file

Commit:
Updated test

Commit:
Moved plugin api to plugins module

aduth commented on docs/extensibility.md:
```diff
@@ -120,38 +120,60 @@ wp.hooks.addFilter(
+    </Fragment>
+);
+
+wp.editPost.__experimental.registerPlugin( 'plugin-names', {
```
`wp.editPost` -> `wp.plugins`

aduth commented on plugins/README.md:
```diff
@@ -0,0 +1,58 @@
+Edit post API
```
Inaccurate heading.

aduth commented on plugins/api/plugin.js:
```diff
@@ -0,0 +1,122 @@
+import { isFunction } from 'lodash';
+
+/**
+ * Singleton class for registering plugins.
```
Why did you choose a singleton pattern? Why the divergence from blocks registration?

Do you plan to follow-up with necessary changes to reinstate consistency if this is the path we go?

aduth replied:
We've also pondered on at least one occasion whether the data backing of blocks should just be `@wordpress/data`. Adding yet another approach into the mix is worsening this consistency.

(If it was data-backed, at least I'd be more comfortable with `isPluginSidebarOpened` operating on the registry, since at that point it's more like composing selectors from other stores)

aduth commented on plugins/api/plugin.js:
```diff
@@ -0,0 +1,122 @@
+ */
+import { applyFilters } from '@wordpress/hooks';
+
+/* External dependencies */
```
Multi-line comment.

aduth commented on plugins/api/index.js:
```diff
@@ -0,0 +1,15 @@
+
+const __experimental = {
+	registerPlugin: registry.registerPlugin,
+	getRegisteredUIComponent: registry.getRegisteredUIComponent,
```
Do we have no way to just retrieve the settings of a plugin? A `getPlugin` method is what I'd expect, at which point a `getRegisteredUIComponent` may not even be necessary.

aduth commented on plugins/components/plugin-area/index.js:
```diff
@@ -0,0 +1,42 @@
+		const pluginRegistry = PluginRegistry.getInstance();
+
+		return (
+			<div id="plugin-fills" style={ { display: 'none' } }>
```
Same feedback applies: https://github.com/WordPress/gutenberg/pull/5430#discussion_r173636497

aduth commented on plugins/components/plugin-area/index.js:
```diff
@@ -0,0 +1,42 @@
+
+		return (
+			<div id="plugin-fills" style={ { display: 'none' } }>
+				{ map( pluginRegistry.plugins, plugin => {
```
Should we be accessing properties of this singleton, or should this be a method from the API?

aduth commented on test/unit/setup-wp-aliases.js:
```diff
@@ -16,6 +16,7 @@ global.wp = {
 	'data',
 	'edit-post',
 	'viewport',
+	'plugins',
```
This also needs to be added to the set in `bin/build-plugin-zip.sh` or it won't be included in the distributed plugin zip.

Xyfi replied:
Done

aduth commented on edit-post/components/plugin-sidebar/sidebar-layout.js:
```diff
@@ -0,0 +1,31 @@
+const SidebarLayout = ( { onClose, title, children } ) => {
+	return (
+		<div
+			className="edit-post-plugin-sidebar"
```
https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

Not entirely clear in the documentation, but this is not the default export of the `index.js`, so is subject to the same rule "for any descendent of the top-level element".

aduth commented on edit-post/components/plugin-sidebar/style.scss:
```diff
@@ -0,0 +1,50 @@
+.edit-post-plugin-sidebar {
+  position: fixed;
```
Tabs, not spaces.

aduth commented on edit-post/components/plugin-sidebar/index.js:
```diff
@@ -0,0 +1,104 @@
+		if ( ! this.namespacedName ) {
+			this.namespacedName = `${ this.props.pluginContext.namespace }/${ this.props.name }`;
+		}
+		this.props.pluginContext.registerUIComponent( this.namespacedName, 'sidebar' );
```
I'm a bit confused by this. This seems to be recreating the same behavior for which Slot/Fill exists to provide. Namely, if we want to register a component for use elsewhere, it simply renders itself into a given fill. Why is this not occupying the general "Sidebar" fill.

Seems like PluginSidebar would both fill the generic sidebar, and also provide its own slot for plugins to fill it.

Xyfi replied:
The problem is that I need a way to check whether an id for a sidebar corresponds to an id for a registered plugin sidebar.

Xyfi replied:
In the `isPluginSidebarOpened` selector to be specific.

aduth commented on edit-post/components/plugin-sidebar/index.js:
```diff
@@ -0,0 +1,104 @@
+	withPluginContext,
+] )( PluginSidebar );
+
+export { PluginSidebarFill, PluginSidebarSlot as PluginSidebar };
```
This feedback still applies: https://github.com/WordPress/gutenberg/pull/5430#discussion_r173636421

aduth commented on edit-post/components/layout/index.js:
```diff
@@ -81,8 +82,9 @@ function Layout( {
-			{ pluginSidebarOpened && <PluginsPanel /> }
+			{ pluginSidebarOpened && <PluginSidebar /> }
 			<Popover.Slot />
+			<PluginArea />
```
If this is truly "PluginArea", why is it specific to edit-post ? Why is it imported as PluginSidebar? Isn't the general pattern that we'd render this area once on every screen?

aduth commented on edit-post/components/plugin-sidebar/sidebar-layout.js:
```diff
@@ -0,0 +1,31 @@
+import './style.scss';
```
Internal dependencies.

Xyfi replied:
Done

Commit:
Updated documentation

Commit:
Converted spaces to tabs in style.scss file

Commit:
Comply to coding style conventions regarding css in plugin-sidebar

Commit:
Added slot to pluginsidebar fill for consistency

Commit:
Moved styles.scss under Internal dependencies

Commit:
removed unnecessary id from hidden div for plugin fills

Commit:
Removed getPluginUIComponent API function in favor of 'plugin-sidebar/'-prefix in sidebar identifier

Commit:
Don't throw error on invalid PluginSidebar props

aduth committed:
Plugins: Refactor plugin UI registration as Slot/Fill

aduth commented:
Take a look at 6540b05dc4a8fa84cc6fb6f1436bd6ccb52f6ca8 as what I had in mind. Let me know what you think.

Specifically, it's leveraging Slot/Fills, generating the name of the intended visible plugin's sidebar as the name of the slot, so only it would be shown.

aduth committed:
Plugins: Remove unused code

aduth committed:
Plugins: Represent plugin render as the component it is

Use createElement

aduth committed:
Plugins: Revert plugin sidebar slot name to PluginSidebar

Apparently the convention we established (see: Popover)

Xyfi commented on plugins/api/index.js:
```diff
@@ -1,15 +1,106 @@
 
-const registry = PluginRegistry.getInstance();
+	settings.name = name;
+	settings.sidebar = {};
```
This doesn't do anything

abotteram reviewed (commented):
A few points I would like to address in regards to @aduth's changes:
- It's allowed to have 2 sidebars with the same name to be rendered, without any feedback why only one is showing.
- It's not enforcing the plugin-sidebar naming conventions.

I still think it's wise to wrap the plugin sidebars in a separate error boundary, because we should protect the editor from crashing because of plugin code as much as possible.

aduth commented:
I don't disagree, but those things are not strictly necessary for an MVP.

I do have thoughts on them though, if you'd like a few more rounds of back-and-forth 😉

aduth commented on edit-post/store/selectors.js:
```diff
@@ -34,13 +34,8 @@ export function isEditorSidebarOpened( state ) {
-
-	return activeGeneralSidebar.startsWith( 'plugin-sidebar/' );
+	const activeGeneralSidebar = getActiveGeneralSidebarName( state );
+	return !! activeGeneralSidebar && ! isEditorSidebarOpened( state );
```
It's the nice thing about a selector is that it abstracts away these underlying implementation-detail concerns.

aduth committed:
Plugins: Remove unused error boundary styles

aduth committed:
Plugin: Remove unused sidebar settings

aduth reviewed (changes requested):
One more thing: The sidebar doesn't look great on mobile. It appears the default sidebars apply a 100% width in a separate selector that this is not benefitting from.

Aside: We need to consolidate our concept of what a sidebar is, so we're not addressing each individually like this.

atimmer committed:
Fix plugin sidebar with on mobile

The className for the plugin sidebar was changed in 644b7e83c. But
this was missed in one location. This commit changes the class there
too.

atimmer committed:
Merge branch 'master' into add/slot-fill-sidebar-api

atimmer committed:
Merge branch 'master' into add/slot-fill-sidebar-api

aduth reviewed (approved):
👍

mcsf commented on plugins/README.md:
```diff
@@ -0,0 +1,58 @@
+    </Fragment>
+);
+
+wp.plugins.__experimental.registerPlugin( 'plugin-names', {
```
This is outdated, `wp.plugins.registerPlugin` is available. (PR description is also outdated.)

aduth replied:
>This is outdated, `wp.plugins.registerPlugin` is available. (PR description is also outdated.)

I missed this. I assumed we were still exporting as `__experimental`. I'm fine with "unexperimental"-izing it at this point.

aduth replied:
>This is outdated, `wp.plugins.registerPlugin` is available. (PR description is also outdated.)

Updated in 853655c.

mcsf commented on bin/build-plugin-zip.sh:
```diff
@@ -99,7 +99,7 @@ zip -r gutenberg.zip \
 	post-content.js \
 	$vendor_scripts \
-	{blocks,components,date,editor,element,hooks,i18n,data,utils,edit-post,viewport}/build/*.{js,map} \
+	{blocks,components,date,editor,element,hooks,i18n,data,utils,edit-post,viewport,plugins}/build/*.{js,map} \
```
Tangent: Perhaps by reading `webpack.config.js`, it'd be nice if we didn't have to manually keep the build script, the ESLint config, etc. in sync.

aduth replied:
>Tangent: Perhaps by reading webpack.config.js, it'd be nice if we didn't have to manually keep the build script, the ESLint config, etc. in sync.

Yes, this has become an increasingly frequent source of friction.

mcsf replied:
--> #5664, for starters

aduth committed:
Documentation: Remove experimental flag from plugins

aduth committed:
Testing: Include plugins folder in ESLint import rule

aduth commented on edit-post/components/layout/index.js:
```diff
@@ -19,6 +19,7 @@ import {
 } from '@wordpress/editor';
 import { withDispatch, withSelect } from '@wordpress/data';
 import { compose } from '@wordpress/element';
+import { PluginArea } from '@wordpress/plugins';
```
>Should we add an entry to `.eslintrc.js`?

Added in a1de97b.

Xyfi commented:
@mcsf Completely agree, but these are tooling issue. Have you created issues for the points you are bringing up?

atimmer reviewed (approved):
Looks like the tooling issues are resolved (or worked on). Giving my own 👍 here and then merging.
